// httprequest/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var NetworkError = /** @class */ (function (_super) {
    __extends(NetworkError, _super);
    function NetworkError(method, url) {
        return _super.call(this, "NetworkError with request: " + method + " " + url) || this;
    }
    return NetworkError;
}(Error));
exports.NetworkError = NetworkError;
/**
Example:

new Request('POST', '/api/reservations').sendData({venue_id: 100}, (err, response) => {
  if (err) throw err
  console.log('got response', res)
})
*/
var Request = /** @class */ (function () {
    /**
    XMLHttpRequest doesn't expose method and url after setting them, so we need
    to keep track of them in the Request instance for error reporting purposes.
    */
    function Request(method, url, responseType) {
        if (responseType === void 0) { responseType = ''; }
        var _this = this;
        this.method = method;
        this.url = url;
        this.headers = [];
        this.xhr = new XMLHttpRequest();
        this.xhr.responseType = responseType;
        this.xhr.onreadystatechange = function (event) {
            var readyState = _this.xhr.readyState;
            if (readyState == 2) { // HEADERS_RECEIVED
                var content_type = _this.xhr.getResponseHeader('content-type');
                if (content_type.indexOf('application/json') === 0) {
                    _this.xhr.responseType = 'json';
                }
                else if (content_type.indexOf('text/xml') === 0) {
                    _this.xhr.responseType = 'document';
                }
            }
        };
        this.xhr.onerror = function (event) {
            _this.callback(new NetworkError(method, url));
        };
        // onload is better than onreadystatechange() { if (readyState == 4) ... }
        //   since onload will not be called if there is an error.
        this.xhr.onload = function (event) {
            if (_this.xhr.status >= 400) {
                var error = new Error(_this.xhr.response);
                return _this.callback(error);
            }
            _this.callback(null, _this.xhr.response);
        };
    }
    /**
    Add a header-value pair to be set on the XMLHttpRequest once it is opened.
    */
    Request.prototype.addHeader = function (header, value) {
        this.headers.push([header, value]);
        return this;
    };
    Request.prototype.send = function (callback) {
        return this.sendData(undefined, callback);
    };
    Request.prototype.sendJSON = function (object, callback) {
        this.headers.push(['Content-Type', 'application/json']);
        return this.sendData(JSON.stringify(object), callback);
    };
    Request.prototype.sendData = function (data, callback) {
        var _this = this;
        this.callback = callback;
        // delay opening until we actually need to so that custom event listeners
        // can be added by the user
        this.xhr.open(this.method, this.url);
        this.headers.forEach(function (_a) {
            var header = _a[0], value = _a[1];
            return _this.xhr.setRequestHeader(header, value);
        });
        try {
            // this might raise an error without even trying the server if we break
            // some kind of cross-origin request rule.
            this.xhr.send(data);
        }
        catch (exc) {
            setTimeout(function () { return callback(exc); }, 0);
        }
        return this;
    };
    return Request;
}());
exports.Request = Request;
